README.md
# [mamma](http://supergiovane.tk/#/mamma)

[![NPM version](https://img.shields.io/npm/v/mamma.svg)](https://www.npmjs.com/package/mamma)
[![Linux Status](https://img.shields.io/travis/hex7c0/mamma.svg?label=linux)](https://travis-ci.org/hex7c0/mamma)
[![Dependency Status](https://img.shields.io/david/hex7c0/mamma.svg)](https://david-dm.org/hex7c0/mamma)
[![Coveralls](https://img.shields.io/coveralls/hex7c0/mamma.svg)](https://coveralls.io/r/hex7c0/mamma)

Offers baby monitors which will provide peace of mind to the new mother.
Open TCP or Unix domain socket and listen child [KA](https://en.wikipedia.org/wiki/Keepalive#TCP_keepalive) keepalive noises.
Alert (callback) when `close` event happen

## Installation

Install through NPM

```bash
npm install mamma
```
or
```bash
git clone git://github.com/hex7c0/mamma.git
```

## API

inside nodejs project
```js
var server = require('mamma').createServer;

var client = require('mamma').createClient;

var bind = require('mamma').createBinding; // bind to existing TCP server
```

### createServer(listen [, options])

#### listen

 - `listen`- **Number | String** Number for `TCP port`, String (path) for `sock` *(default "required")*

#### options

 - `keepalive` - **Number** Set initialDelay (in milliseconds) to set the delay between the last data packet received and the first keepalive probe *(default "2000")*
 - `callback`- **Function** Set callback when `close` event happen (return had_error,id) *(default "disabled")*
 - `http` - **Object** Enable http interface *(default "disabled")*
   - `port` - **Number** Set Web port *(default "3000")*
   - `host` - **String** Set Web host *(default "127.0.0.1")*
 - `https` - **Object** Enable https interface [extra options](http://nodejs.org/api/https.html#https_https_createserver_options_requestlistener) *(default "disabled")*
   - `port` - **Number** Set Web port *(default "3000")*
   - `host` - **String** Set Web host *(default "127.0.0.1")*
 - `host` - **String** If the host is omitted, the server will accept connections directed to any IPv4 address (TCP only) *(default "omitted")*

### createClient(connect, id [, options])

#### connect

 - `connect`- **Object** [createConnection](http://nodejs.org/api/net.html#net_net_createconnection_options_connectionlistener) *(default "required")*

#### id

 - `id`- **String** Child id for callback *(default "required")*

#### options

 - `autoReconnect` - **Boolean** Set autoReconnect flag *(default "true")*
 - `maxRetries`- **Number** Set maxRetries for autoReconnect logic (true for unlimited) *(default "true")*
 - `delay` - **Number** Set millisecond delay from each try *(default "2000")*

### createBinding(connect, id [, options])

#### connect

 - `connect`- **Object** [createConnection](http://nodejs.org/api/net.html#net_net_createconnection_options_connectionlistener) *(default "required")*

#### id

 - `id`- **String** Child id for callback *(default "required")*

#### options

 - `autoReconnect` - **Boolean** Set autoReconnect flag *(default "true")*
 - `maxRetries`- **Number** Set maxRetries for autoReconnect logic (true for unlimited) *(default "true")*
 - `delay` - **Number** Set millisecond delay from each try *(default "2000")*
 - `keepalive` - **Number** Set initialDelay (in milliseconds) to set the delay between the last data packet received and the first keepalive probe *(default "2000")*
 - `callback`- **Function** Set callback `close` event happen (return had_error,id) *(default "disabled")*

## Examples

Take a look at my [examples](examples)

### [License GPLv3](LICENSE)


History.md
v0.2.3 / 2015-12-03
==================

  * Test `node`@5

v0.2.2 / 2015-10-15
==================

  * Test `node`@4

v0.2.1 / 2015-08-15
==================

  * SPDX license

v0.2.0 / 2015-03-19
==================

  * `iojs` test
  * `coveralls` test

v0.1.3 / 2015-01-18
==================

  * Fix regression bug

v0.1.1 / 2015-01-18
==================

  * Support for TCP host
  * Add "host" (options)

v0.1.0 / 2015-01-17
==================

  * Add createBinding method
  * Add autoReconnect for createClient and createBinding

v0.0.4 / 2015-01-15
==================

  * Remove "console" (options)
  * Remove `error` event listener

v0.0.3 / 2015-01-15
==================

  * Little code minimization

v0.0.2 / 2015-01-15
==================

  * Add "console" (options)

v0.0.1 / 2015-01-15
==================

  * Project start


test/tcp.js
'use strict';
/**
 * @file tcp test
 * @module mamma
 * @subpackage test
 * @version 0.0.1
 * @author hex7c0 <hex7c0@gmail.com>
 * @copyright hex7c0 2014
 * @license GPLv3
 */

/*
 * initialize module
 */
var mamma = require('..');
var assert = require('assert');

/*
 * test module
 */
describe('tcp', function() {

  var port = 3000;

  describe('server', function() {

    it('should return an Error because no "listen"', function(done) {

      try {
        mamma.createServer();
      } catch (err) {
        assert.equal(err.message, 'listen required');
        done();
      }
    });
    it('should return Server obj', function(done) {

      var m = mamma.createServer(port);
      assert.ok(m.close);
      m.close();
      done();
    });
  });

  describe('client', function() {

    it('should return an Error because no "connect"', function(done) {

      try {
        mamma.createClient();
      } catch (err) {
        assert.equal(err.message, 'connect required');
        done();
      }
    });
    it('should return an Error because "connect" isn\'t an Object',
      function(done) {

        try {
          mamma.createClient('ciao');
        } catch (err) {
          assert.equal(err.message, 'connect required');
          done();
        }
      });
    it('should return an Error because no "id"', function(done) {

      try {
        mamma.createClient({
          port: port
        });
      } catch (err) {
        assert.equal(err.message, 'id required');
        done();
      }
    });
    it('should return Client obj and ECONNREFUSED', function(done) {

      var m = mamma.createClient({
        port: port
      }, 'ciao', {
        autoReconnect: false
      });
      assert.ok(m.end);
      m.on('error', function(err) {

        assert.ok(/connect ECONNREFUSED/.test(err.message));
        done();
      });
    });
  });

  describe('run', function() {

    it('should test client end', function(done) {

      var s = mamma.createServer(port);
      var c = mamma.createClient({
        port: port
      }, 'should test client end', {
        autoReconnect: false
      });
      setTimeout(function() {

        c.end();
        s.close();
        done();
      }, 1000);
    });
    it('should test server end', function(done) {

      var s = mamma.createServer(port);
      var c = mamma.createClient({
        port: port
      }, 'should test server end', {
        autoReconnect: false
      });
      setTimeout(function() {

        s.close();
        assert.equal(c.writable, true);
        c.end();
        done();
      }, 1000);
    });
    it('should test keepalive probe with 10000 timeout', function(done) {

      this.timeout(10000);
      var s = mamma.createServer(port);
      var c = mamma.createClient({
        port: port
      }, 'should test keepalive probe with 10000 timeout', {
        autoReconnect: false
      });
      setTimeout(function() {

        assert.equal(c.destroyed, false);
        s.getConnections(function(err, count) {

          assert.equal(count, 1);
          c.end();
          s.close();
          done();
        });
      }, 9900);
    });
    it('should test multiple client', function(done) {

      var s = mamma.createServer(port);
      var c1 = mamma.createClient({
        port: port
      }, 'multiple1', {
        autoReconnect: false
      });
      var c2 = mamma.createClient({
        port: port
      }, 'multiple2', {
        autoReconnect: false
      });
      setTimeout(function() {

        s.getConnections(function(err, count) {

          assert.equal(count, 2);
          c1.end();
          c2.end();
          s.close();
          done();
        });
      }, 500);
    });
  });
});


test/http.js
'use strict';
/**
 * @file http test
 * @module mamma
 * @subpackage test
 * @version 0.0.1
 * @author hex7c0 <hex7c0@gmail.com>
 * @copyright hex7c0 2014
 * @license GPLv3
 */

/*
 * initialize module
 */
var mamma = require('..');
var assert = require('assert');
var request = require('superagent');

/*
 * test module
 */
describe('http', function() {

  var port = 3000;
  var uri = 'http://127.0.0.1:3001/';
  var server;

  before(function(done) {

    server = mamma.createServer(port, {
      http: {
        port: port + 1
      }
    });
    done();
  });

  it('should test empty page', function(done) {

    request.get(uri).end(function(err, res) {

      assert.equal(err, null);
      assert.equal(res.statusCode, 200);
      assert.equal(res.text, '{}');
      done();
    });
  });
  it('should test / page with 1 child', function(done) {

    var c = mamma.createClient({
      port: port
    }, 'child 1', {
      autoReconnect: false
    });
    setTimeout(function() {

      request.get(uri).end(function(err, res) {

        assert.equal(err, null);
        assert.equal(res.statusCode, 200);
        var j = JSON.parse(res.text);
        assert.equal(j['child 1'].state, 'open');

        c.end();
        setTimeout(function() {

          request.get(uri).end(function(err, res) {

            assert.equal(err, null);
            assert.equal(res.statusCode, 200);
            var j = JSON.parse(res.text);
            assert.equal(j['child 1'].state, 'closed');
            done();
          });
        }, 500);

      }, 500);
    });
  });
  it('should test child page', function(done) {

    request.get(uri + 'child 1').end(function(err, res) {

      assert.equal(err, null);
      assert.equal(res.statusCode, 200);
      var j = JSON.parse(res.text);
      assert.equal(j.state, 'closed');
      done();
    });
  });
  it('should test another child page', function(done) {

    request.get(uri + 'ciao222').end(function(err, res) {

      assert.notEqual(err, null);
      assert.equal(err.response.statusCode, 404);
      assert.equal(res.statusCode, 404);
      assert.equal(res.text, '');
      done();
    });
  });
  it('should close server', function(done) {

    server.close();
    done();
  });
});


test/https.js
'use strict';
/**
 * @file http test
 * @module mamma
 * @subpackage test
 * @version 0.0.1
 * @author hex7c0 <hex7c0@gmail.com>
 * @copyright hex7c0 2014
 * @license GPLv3
 */

/*
 * initialize module
 */
var mamma = require('..');
var assert = require('assert');
var request = require('superagent');
var fs = require('fs');
process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0'; // force

/*
 * test module
 */
describe('https', function() {

  var port = 3000;
  var uri = 'https://127.0.0.1:3002/';
  var server;

  before(function(done) {

    server = mamma.createServer(port, {
      https: {
        port: port + 2,
        key: fs.readFileSync(__dirname + '/test_key.pem'),
        cert: fs.readFileSync(__dirname + '/test_cert.pem')
      }
    });
    done();
  });

  it('should test empty page', function(done) {

    request.get(uri).end(function(err, res) {

      assert.equal(err, null);
      assert.equal(res.statusCode, 200);
      assert.equal(res.text, '{}');
      done();
    });
  });
  it('should test / page with 1 child', function(done) {

    var c = mamma.createClient({
      port: port
    }, 'child 1', {
      autoReconnect: false
    });
    setTimeout(function() {

      request.get(uri).end(function(err, res) {

        assert.equal(err, null);
        assert.equal(res.statusCode, 200);
        var j = JSON.parse(res.text);
        assert.equal(j['child 1'].state, 'open');

        c.end();
        setTimeout(function() {

          request.get(uri).end(function(err, res) {

            assert.equal(err, null);
            assert.equal(res.statusCode, 200);
            var j = JSON.parse(res.text);
            assert.equal(j['child 1'].state, 'closed');
            done();
          });
        }, 500);

      }, 500);
    });
  });
  it('should test child page', function(done) {

    request.get(uri + 'child 1').end(function(err, res) {

      assert.equal(err, null);
      assert.equal(res.statusCode, 200);
      var j = JSON.parse(res.text);
      assert.equal(j.state, 'closed');
      done();
    });
  });
  it('should test another child page', function(done) {

    request.get(uri + 'ciao222').end(function(err, res) {

      assert.notEqual(err, null);
      assert.equal(err.response.statusCode, 404);
      assert.equal(res.statusCode, 404);
      assert.equal(res.text, '');
      done();
    });
  });
  it('should close server', function(done) {

    server.close();
    done();
  });
});


index.js
'use strict';
/**
 * @file mamma main
 * @module mamma
 * @subpackage main
 * @version 0.2.0
 * @author hex7c0 <hex7c0@gmail.com>
 * @copyright hex7c0 2014
 * @license GPLv3
 */

/*
 * initialize module
 */
var net = require('net');

/*
 * functions
 */
/**
 * createServer
 * 
 * @exports createServer
 * @function createServer
 * @param {Number|String} listen - listen type
 * @param {Object} [opt] - various options. Check README.md
 * @return {Object}
 */
function createServer(listen, opt) {

  if (!listen) {
    throw new TypeError('listen required');
  }
  var options = opt || Object.create(null);
  var my = {
    keepalive: Number(options.keepalive) || 2000,
    callback: typeof options.callback === 'function' ? options.callback : false,
    http: false,
    https: false
  };
  if (typeof options.http === 'object') {
    my.http = {
      port: Number(options.http.port) || 3000,
      host: String(options.http.host || '127.0.0.1')
    };
  }
  if (typeof options.https === 'object') {
    my.https = options.https;
    my.https.port = Number(options.https.port) || 3000;
    my.https.host = String(options.https.host || '127.0.0.1');
  }

  var hosts = Object.create(null);
  var server = net.createServer(function(sock) {

    sock.setKeepAlive(true, my.keepalive); // https://en.wikipedia.org/wiki/Keepalive
    sock.on('data', function(buff) {

      if (sock._id) {
        hosts[sock._id] = {
          state: sock.readyState,
          beat: Date.now()
        };
      } else if (buff.length > 1) { // set client id
        var _id = String(buff);
        hosts[_id] = {
          state: sock.readyState,
          beat: Date.now()
        };
        sock._id = _id;
      }
      return;

    }).on('close', function(had_error) {

      if (sock._id) {
        hosts[sock._id] = {
          state: sock.readyState,
          beat: Date.now()
        };
      }
      // error hook [with client id]
      return my.callback !== false ? my.callback(had_error, sock._id) : null;
    });
  });

  if (options.host && typeof listen === 'number') { // tpc only
    server.listen(listen, options.host);
  } else {
    server.listen(listen);
  }

  if (my.http || my.https) {
    var web = function(req, res) {

      var out = '';
      res.setHeader('Content-Type', 'application/json; charset=utf-8');
      var url = decodeURI(req.url);
      url = url.substr(1, url.length);
      if (url === '') {
        out = JSON.stringify(hosts);
      } else if (hosts[url] !== undefined) {
        out = JSON.stringify(hosts[url]);
      } else {
        res.statusCode = 404;
      }
      return res.end(out);
    };
    if (my.http) {
      require('http').createServer(web).listen(my.http.port, my.http.host);
    }
    if (my.https) {
      require('https').createServer(my.https, web).listen(my.https.port,
        my.https.host);
    }
  }

  server._hosts = hosts; // hosts reference
  return server;
}
module.exports.createServer = createServer;

/**
 * createClient
 * 
 * @exports createClient
 * @function createClient
 * @see http://nodejs.org/api/net.html#net_net_createconnection_options_connectionlistener
 * @param {Object} connect - connect options
 * @param {String} id - child id
 * @param {Object} [opt] - various options. Check README.md
 * @return {Object}
 */
function createClient(connect, id, opt) {

  if (!connect || typeof connect !== 'object') {
    throw new TypeError('connect required');
  } else if (!id) {
    throw new TypeError('id required');
  }
  var options = opt || Object.create(null);
  var my = {
    autoReconnect: options.autoReconnect === false ? false : true,
    maxRetries: Number(options.maxRetries) || true,
    delay: Number(options.delay) || 2000
  };

  function autoReconnect() {

    if (my.maxRetries === true || --maxRetries > 0) {
      setTimeout(function() {

        if (connect.port) { // tpc
          client.connect(connect.port, connect.host);
        } else { // unix socket
          client.connect(connect.path);
        }
        return;
      }, my.delay);
    }
    return;
  }

  var client = net.createConnection(connect).on('connect', function() {

    return client.write(String(id));
  });
  if (my.autoReconnect) {
    client.on('close', autoReconnect);
  }
  client.close = function() {

    client.removeListener('close', autoReconnect);
    return client.end();
  };

  return client;
}
module.exports.createClient = createClient;

/**
 * createBinding
 * 
 * @exports createBinding
 * @function createBinding
 * @see http://nodejs.org/api/net.html#net_net_createconnection_options_connectionlistener
 * @param {Object} connect - connect options
 * @param {String} id - child id
 * @param {Object} [opt] - various options. Check README.md
 * @return {Object}
 */
function createBinding(connect, id, opt) {

  if (!connect || typeof connect !== 'object') {
    throw new TypeError('connect required');
  } else if (!id) {
    throw new TypeError('id required');
  }
  var options = opt || Object.create(null);
  var my = {
    autoReconnect: options.autoReconnect === false ? false : true,
    maxRetries: Number(options.maxRetries) || true,
    delay: Number(options.delay) || 2000,
    keepalive: Number(options.keepalive) || 2000,
    callback: typeof options.callback === 'function' ? options.callback : false,
  };

  function closed(had_error) {

    if (my.autoReconnect === true) {
      if (my.maxRetries === true || --maxRetries > 0) {
        setTimeout(function() {

          if (connect.port) { // tpc
            client.connect(connect.port, connect.host);
          } else { // unix socket
            client.connect(connect.path);
          }
          return;
        }, my.delay);
      }
    }
    // error hook [with client id]
    return my.callback !== false ? my.callback(had_error, id) : null;
  }

  var client = net.createConnection(connect);
  client.on('close', closed);
  client.setKeepAlive(true, my.keepalive);
  client.close = function() {

    client.removeListener('close', closed);
    return client.end();
  };

  return client;
}
module.exports.createBinding = createBinding;
